Shuffleboard PID tuning widgets

FireShredder24 opened the pull request:
Add Shuffleboard PID widgets for drivetrain and shooter
Add more status widgets for shooter hood, shooter, and drivetrain

FireShredder24 committed:
Replace SmartDashboard widgets

Replace existing SmartDashboard status widgets with the appropriate Shuffleboard widgets

FireShredder24 committed:
Add PID Widgets for Shooter

FireShredder24 committed:
Add PID Widgets for Drive

Add constant updater method to periodic() of both FalconDriveSubsystem and ShooterSubsystem

FireShredder24 committed:
Remove duplicates, change speed widget to graph

terbos90803 committed:
Meeting changes

terbos90803 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -60,33 +60,36 @@
-  private final JoystickButton intakeRollerFwdButton = new JoystickButton(f310, LogitechF310.Button.kBumperLeft.value);
-  private final JoystickButton intakeRollerRvsButton = new JoystickButton(f310, LogitechF310.Button.kBumperRight.value);
-  private final JoystickButton intakePistonToggle = new JoystickButton(f310, LogitechF310.Button.kB.value);
+  private final JoystickButton intakeRollerFwdButton = new JoystickButton(f310, LogitechF310.Button.kBumperLeft.value); // Left
```
Does the comment add anything here?  Is kBumperLeft not readable?  Adding a comment also adds maintenance.

Same feedback for all the comments added here.

FireShredder24 replied:
In hindsight, it didn't help much.  Will revert after homework is done.

terbos90803 commented on src/main/java/frc/robot/commands/DriveCommand.java:
```diff
@@ -35,7 +35,7 @@ public void initialize() {
   public void execute() {
     m_subsystem.drive(m_conditioner.conditionMove(m_controller.getMoveRequest()),
-        m_conditioner.conditionTurn(m_controller.getTurnRequest()));
+        m_conditioner.conditionTurn(-m_controller.getTurnRequest()));
```
Why was this needed now?  It was working correctly before.

FireShredder24 replied:
Well, turning was inverted.  I don't know why, but it was.  I'd guess that some tweaking to Conditioners over the summer did that.

terbos90803 replied:
It's much more important to understand why than to just patch until working.  Doing that often leads to a big mess.
For instance, it is normal to invert the raw Y axis because of the long history of video games on legacy computers where Y is positive down on the screen.
It is not normal to invert X, so I'd look into this a lot more closely.

FireShredder24 replied:
Jordan fixed this by adding a FunctionalConditioner to the ComposedConditioner that is passed to DriveCommand in RobotContainer.  It reverses turnRequest while leaving moveRequest untouched.

terbos90803 commented on src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:
```diff
@@ -58,31 +59,44 @@
 
+  private NetworkTableEntry leftDriveVelocity;
+  private NetworkTableEntry rightDriveVelocity;
+  private NetworkTableEntry kPSlider;
```
There used to be a combined element to show all the PID coeffs together.  Is that still around?

FireShredder24 replied:
There is, but it doesn't work with Talon FX(or Spark MAX), since neither extends the WPILib PIDController class.

terbos90803 replied:
I thought Jordan wrote a wrapper that didn't rely on that class.  Or am I mis-remembering.

FireShredder24 replied:
According to the Java API, the widget kPIDController ([here](https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/shuffleboard/BuiltInWidgets.html#kPIDController)) relies on a deprecated version of PIDController.  It's weird.  For now, I've just been using separate widgets, organizing them into lists.  It's not very screen-space-efficient, but it works.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -103,7 +109,8 @@ private boolean hasReliableZero() {
 
   private boolean isFullyDeployed() {
-    return Math.abs(getHoodPos() - currSetpoint) < MoPrefs.getShooterHoodPositionTolerance();
+    return MoPrefs.getShooterHoodSetpoint() - MoPrefs.getShooterHoodPositionTolerance() < getHoodPos()
```
Making tolerance comparisons should always be in the form:
Math.abs(setpoint - currPos) < tolerance

Why did this get changed?

FireShredder24 replied:
I was having a lot of trouble with the it, so I messed around with the logic.  Eventually, it turned out that getHoodPos() wasn't actually polling the encoder.  I just didn't change it back because we moved on to tuning the shooter flywheel PID.  I'll fix it once I finish my homework.

terbos90803 replied:
Another case of needing better understanding before hacking.
I get that if you aren't familiar with the standard tolerance pattern that it may have seemed odd, but now it seems odd and difficult to read for the rest of us.

As a general rule, any time you have to work to understand a bit of code, that's a screaming flag that it deserves a comment block to explain what's going on and why.  This is doubly true if you need to change something to work around a bug.  When the real bug gets fixed, it needs to be easy to find any workarounds that will no longer behave as expected and fix them too.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -165,19 +198,38 @@ public void purge() {
+  }
+
+  private void updatePIDConstants() {
+    shooterPIDLeft.setP(kP, 0);
```
You can't have two motors mechanically connected and also PID them separately.  The PID has to be on one of the motors and the other motor follows the first one.

FireShredder24 committed:
Remove unnecessary comments

FireShredder24 committed:
Revert hood verification logic

Also, remove unnecessary setHoodPos(double)

FireShredder24 committed:
Remove unnecessary PID settings from shooter

Add more shooter hood comments

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -91,19 +98,24 @@ public void stowHood() {
 
   public double getHoodPos() {
-    return hoodPos;
+    return hoodEncoder.getPosition();
```
Just out of curiosity, why did this move here?  It used to be in periodic().  Unless we truly need to get the hood position more frequently than regular period, it's better to read it once and cache the result.

FireShredder24 replied:
Fixed

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -112,21 +124,23 @@ public boolean isHoodReady() {
 
   @Override
   public void periodic() {
-    hoodPos = hoodEncoder.getPosition();
```
This line needs to come back.

FireShredder24 replied:
It has

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -125,16 +126,46 @@ public ShooterSubsystem(final ShooterHoodSubsystem shooterHood, ShuffleboardTab
+    shooterSetpoint.addListener(notice -> setpoint = notice.value.getDouble(),
+        EntryListenerFlags.kNew | EntryListenerFlags.kUpdate);
+
+    kPSlider = tab.addPersistent("Shooter kP", 5e-5).withWidget(BuiltInWidgets.kTextView).getEntry();
```
There should be no hardcoded constants in these calls.  Especially since these are tuning values and already exist elsewhere.  Keeping them in sync is high maintenance.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -125,16 +126,46 @@ public ShooterSubsystem(final ShooterHoodSubsystem shooterHood, ShuffleboardTab
     // deploy hood
     // run gate if both of "" are good
 
-    double pidSetpoint = MoPrefs.getShooterPIDSetpoint();
```
What the heck is going on here?  The setpoint should be a MoPref and not hardcoded (aside from a default).

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -165,19 +196,33 @@ public void purge() {
+    isFlywheelReady.setBoolean(isFlywheelReady());
+    updatePIDConstants();
+
+    setpoint = shooterSetpoint.getDouble(4500);
```
What's going on here?  Why is the setpoint being updated every periodic and there's a hardcoded constant which is repeated multiple times in the code.

jordan-powers committed:
Fix graphs on SmartDashboard

Turns out that the SmartDashboard isn't fond of special characters in the
NetworkTables keys.

jordan-powers committed:
Remove turn request negation from DriveCommand

A negation was added in DriveCommand as a quick fix for an inverted turn
request. The proper way to fix this is to use a DriveConditioner.
This fix reverts the negation in DriveCommand and adds a new
DriveConditioner that inverts the turn request.

This fixes the broken DriveCommandTest unit test.

jordan-powers committed:
ShooterHoodSubsystem: Update hoodPos in periodic()

FireShredder24 committed:
Revert shooter setpoint to a MoPref

Add comment in ControllerBase

Update MAX_FREE_SPEED to match observed maximum free speed

FireShredder24 committed:
Replace shooter PID constants with MoPrefs

FireShredder24 committed:
Remove empty NetworkTableEntries

FireShredder24 committed:
Replace Drive PID widgets with MoPrefs

FireShredder24 committed:
Replace ShooterHood PID Constants with MoPrefs

FireShredder24 committed:
Fix new code smells

terbos90803 commented on src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:
```diff
@@ -185,4 +159,40 @@ public void stop() {
+    double kD = MoPrefs.getDriveKD();
+    int kIZ = (int) MoPrefs.getDriveKIZ();
+    double kFF = MoPrefs.getDriveKFF();
+    leftFront.config_kP(0, kP);
```
I know it adds complexity, but I seem to recall that you shouldn't be writing PID constants every period.  You'll have to figure out how to only write them on changes.  That usually requires caching the old value and only writing on diffs.  Jordan can probably help make this look all fancy so the code doesn't become cumbersome.

jordan-powers replied:
This would be easy if the Preferences class allowed us access to the underlying `NetworkTableEntry`s.  Unfortunately, that is not the case.
Also, references is a final class, so we can't subclass it and give ourselves access.

I can see two options:
1. Re-implement the Preferences class, and add a method to get access to the `NetworkTableEntry`. The Preferences class is pretty simple, so this wouldn't be too difficult. It does seem a bit hacky though.
2. Create some sort of "watcher" class that watches a value, looks for changes, and triggers a lambda when the value changes.  This would be pretty trivial to implement, however it would require polling the Preferences NetworkTable on every loop. This is probably not a big deal, though, and I suspect the NetworkTables are designed to support it.

I'm not sure which option is better.

terbos90803 replied:
I thought MoPrefs already subclassed Preferences.  Apparently, it does not, so it must wrap it instead (composition vs inheritance).

I also thought there was a way to subscribe to NT entries, but if there's not, then having a mechanism in place to detect and notify on changes would certainly be helpful.

FireShredder24 replied:
What if instead of polling Preferences and writing the values to the Talon FXs every scheduler cycle, we only did that once per second?  That would be trivially easy to implement and would basically accomplish the same goal.

terbos90803 replied:
It might help to understand the root problem.
The SparkMax docs say specifically that certain values like the PID settings shouldn't be written frequently.  I suspect the TalonFX has a similar warning.
The reason is because those values are held in a form of non-volatile memory (flash or eeprom) which has a limited number of write cycles in its lifetime.  You want to make sure you only write that when you're changing it.  Writing the same value repeatedly only wears out the storage and shortens the life of the device.

Theoretically, this type of check should be happening in the vendor code between you and the device, but you can never know for sure and if the docs say don't do it, then it's a safer bet to assume they aren't checking for you.

jordan-powers replied:
There is a way to subscribe to Networktable entries. However, the WPILib Preferences class doesn't let us access those entries. Just the values.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -165,19 +139,32 @@ public void purge() {
   }
 
+  private void updatePIDConstants() {
+    shooterPIDRight.setP(MoPrefs.getShooterKP(), 0);
```
Only update on change

terbos90803 commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -86,4 +109,72 @@ public static double getShooterPIDSetpoint() {
     return getDouble("Shoot From Wall Hood Setpoint", SHOOT_FROM_WALL_HOOD_SETPOINT);
   }
+
+  public static double getShooterKP() {
```
Does anyone else find this code pattern a challenge to maintain?
Why do we need specific methods baked into a shared class to get values from a dictionary?
It has always seemed to me that this should be easier.  Specifically, MoPrefs should maintain the dictionary, but not know/care about what's in it.  The clients should be responsible for knowing their own keys.

jordan-powers replied:
Well, MoPrefs is just a wrapper around ` edu.wpi.first.wpilibj.Preferences`. The only functionality it adds is the individual method calls. As far as I recall, these are the reasons for why we chose to create the MoPrefs instead of directly using the Preferences object:
1. To provide explicit default values for each preference
2. To make it very obvious what values are stored in the preferences
3. To prevent errors where typos in a string key ends up getting unexpected values

However, I do agree that this way of achieving that goal is pretty clunky.

Maybe a better way to do this would be to create a `enum MoPrefKeys` and refactor `MoPrefs` to only have one public method which takes a MoPrefKey and returns the value. Java's enums are basically just classes, so we could define a MoPrefKey to hold both the string key and an explicit default value.

That shouldn't be too hard to implement: I'll create a PR.

jordan-powers replied:
See #27

jordan-powers committed:
Add MoPrefsKey enum

MoPrefs are now accessed using the get() method, which takes a MoPrefsKey.
MoPrefsKeys are responsible for tracking the default value of their
associated preference, and are used as keys into the Preferences table.

This improves maintainability: adding a new preference only requires
adding a new entry to the MoPrefsKey enum instead of a whole new method
on MoPrefs.

jordan-powers committed:
Finish conversion to using MoPrefsKey enum

jordan-powers committed:
Merge pull request #27 from momentumfrc/moprefs-enum

Add MoPrefsKey enum

aimiebell commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -47,43 +123,11 @@ public static void setDouble(String key, double value) {
     System.out.format("set pref: %s=%f\n", key, value);
```
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.
